main.go

```
@@ -0,0 +1,136 @@
+package main
+
+import (
+  "io"
+  "bufio"
+  "fmt"
+  "net/http"
+  "encoding/csv"
+  "encoding/json"
+  "math"
+  "strconv"
+  "sort"
+  "os"
+)
+
+type Shop struct {
+  Name string
+  Y float64
+  X float64
+}
+
+type ShopDistance struct {
+  Name string
+  Distance float64
+}
+
+func main() {
+  http.HandleFunc("/needcoffee", needCoffee)
+  http.HandleFunc("/addcoffees", addCoffees)
+
+  http.ListenAndServe(":4000", nil)
+}
+
+func distance(x1, y1, x2, y2 float64) float64 {
+  newX := math.Pow((x1-x2), 2)
+  newY := math.Pow((y1-y2), 2)
+  sum := newX+newY
+  sumSq := math.Sqrt(sum)
+  return (toFixed(sumSq, 4))
+}
+
+func round(num float64) int {
+    return int(num + math.Copysign(0.5, num))
+}
+
+func toFixed(num float64, precision int) float64 {
+    output := math.Pow(10, float64(precision))
+    return float64(round(num * output)) / output
+}
+
+func readShops() []Shop {
+  var shops = []Shop{}
+
+  csvFile, _ := os.Open("data.csv")
+  reader := csv.NewReader(bufio.NewReader(csvFile))
+  for {
+    line, error := reader.Read()
+    if error == io.EOF {
+        break
+    } else if error != nil {
+        fmt.Println(error)
+    }
+    y, _ := strconv.ParseFloat(line[1], 64)
+    x, _ := strconv.ParseFloat(line[2], 64)
+    shops = append(shops, Shop{
+      Name: line[0],
+      Y: y,
+      X: x,
+    })
+  }
+
+  return shops
+}
+
+func needCoffee(w http.ResponseWriter, r *http.Request) {
+
+  shops := readShops()
+
+  xString := r.URL.Query()["x"][0]
+  yString := r.URL.Query()["y"][0]
+
```

should return an error if the expected params aren't provided